File: VPKitDemoObjC/VPKitDemo/ViewController.m
//
//  ViewController.m
//  VPKitDemo
//
//  Created by jonathan on 01/03/2016.
//  Copyright © 2016 jonathan. All rights reserved.
//

#import "ViewController.h"
#import "Common.h"

@import AVFoundation;
@import AVKit;
#import <VPKit/VPKit.h>


@interface ViewController ()
<
  VPKVeepViewerDelegate
, VPKVeepEditorDelegate
, VPKPreviewDelegate
>

@property (nonatomic, strong) VPKVeepViewer* vpViewer;
@property (nonatomic, strong) VPKVeepEditor* vpEditor;
@property (nonatomic, strong) UIImageView* imageButton;
@property (nonatomic, strong) VPKPreview* vpkPreview;
@property (nonatomic, strong) UILabel* consumeLabel;
@property (nonatomic, strong) UILabel* createLabel;
@property (nonatomic, strong) UILabel* titleLabel;

@end

@implementation ViewController

#pragma mark - lazy initialisers



-(VPKPreview*)vpkPreview {
    if (!_vpkPreview) {
        _vpkPreview = [[VPKPreview alloc] init];
        _vpkPreview.translatesAutoresizingMaskIntoConstraints = NO;
    }
    return _vpkPreview;
}

- (UIImageView*)imageButton {
    if (!_imageButton) {
        UIButton* button = [UIButton buttonWithType:UIButtonTypeCustom];
        button.translatesAutoresizingMaskIntoConstraints = NO;
        [button addTarget:self
                   action:@selector(imageViewButtonPushed:)
         forControlEvents:UIControlEventTouchUpInside];
        
        _imageButton = [[UIImageView alloc] initWithFrame:CGRectZero];
        _imageButton.translatesAutoresizingMaskIntoConstraints = NO;
        _imageButton.image = [UIImage imageNamed:@"stock_photo"];
        _imageButton.contentMode = UIViewContentModeScaleAspectFit;
        _imageButton.userInteractionEnabled = YES;
        [_imageButton addSubview:button];
        
        NSArray* formats =
        @[
          @"V:|[button]|",
          @"H:|[button]|"
          ];
        
        for (NSString* format in formats) {
            [_imageButton addConstraints:
             [NSLayoutConstraint constraintsWithVisualFormat:format
                                                     options:0 metrics:@{}
                                                       views:@{@"button":button}]];
        }
    }
    return _imageButton;
}


#pragma mark - viewController lifecycle

- (void)dealloc {
    [self stopListening];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self configureLabels];
    [self configureVpkPreview];
    [self configureImageButton];
    [self configureConstraints];
    [self startListening];
}

#pragma mark - configuration

- (void)configureLabels {
    self.createLabel = [self newLabel:@"Create Veep"];
    self.consumeLabel = [self newLabel:@"Consume Veep"];
    self.titleLabel = [self newLabel:@"VEEPIO SDK Demo"];
}

- (UILabel*)newLabel:(NSString*)text {
    UILabel* label = [[UILabel alloc] init];
    label.translatesAutoresizingMaskIntoConstraints = NO;
    label.textAlignment = NSTextAlignmentCenter;
    label.font = [UIFont systemFontOfSize:14];
    label.text = text;
    [self.view addSubview:label];
    return label;
}

- (void)configureImageButton {
    [self.view addSubview:self.imageButton];
}

- (void)configureVpkPreview {
    [self.view addSubview:self.vpkPreview];
    UIImage* image = [UIImage imageNamed:@"KrispyGlas"];
    image = [[VPKImage alloc] initWithImage:image veepID:@"674"];
    self.vpkPreview.image = image;
    
    /*
     setting the delegate is OPTIONAL - if it is unset, default behaviour handles presenting and dismissing
     */
    
    self.vpkPreview.delegate = self;
}

- (void)configureConstraints {
    NSDictionary* dict = NSDictionaryOfVariableBindings(_titleLabel,_vpkPreview,_consumeLabel,_imageButton,_createLabel);
    CGFloat vMargin = 50;
    NSArray* formats =
    @[
     @"H:|-(vMargin)-[_titleLabel]-(vMargin)-|",
     @"H:|-(vMargin)-[_vpkPreview]-(vMargin)-|",
     @"H:|-(vMargin)-[_consumeLabel]-(vMargin)-|",
     @"H:|-(vMargin)-[_imageButton]-(vMargin)-|",
     @"H:|-(vMargin)-[_createLabel]-(vMargin)-|",
     @"V:|-(38)-[_titleLabel]-(18)-[_vpkPreview]-(4)-[_consumeLabel]-(18)-[_imageButton]-(4)-[_createLabel]"
     ];
    
    
    for (NSString* format in formats) {
        [self.view addConstraints:
         [NSLayoutConstraint constraintsWithVisualFormat:format
                                                 options:0
                                                 metrics:@{@"vMargin":@(vMargin)}
                                                   views:dict]];
    }
    CGSize size = self.vpkPreview.image.size;
    [self alignHeightToWidth:self.vpkPreview ratio: size.height/size.width];
    size = self.imageButton.image.size;
    [self alignHeightToWidth:self.imageButton ratio: size.height/size.width];
    
}

- (void)alignHeightToWidth:(UIView*)view ratio:(CGFloat)ratio{
    [view.superview addConstraint:
     [NSLayoutConstraint constraintWithItem:view
                                  attribute:NSLayoutAttributeHeight
                                  relatedBy:NSLayoutRelationEqual
                                     toItem:view
                                  attribute:NSLayoutAttributeWidth
                                 multiplier:ratio
                                   constant:0]];
}

#pragma mark - interactions


- (void)imageViewButtonPushed:(UIButton*)sender {
    /*
     invoking the VPKVeepEditor
     
     set the editor's transitioning delegate to a custom transitioning object (or nil) to override supplied transition animations

     */
    self.vpEditor = [VPKit editorWithImage:self.imageButton.image
                                  fromView:self.imageButton];
    if (self.vpEditor) {
        self.vpEditor.delegate = self;
        self.vpEditor.modalPresentationStyle = UIModalPresentationOverFullScreen;
        [self presentViewController:self.vpEditor animated:YES completion:nil];
    }
}


#pragma mark - VPKPreview delegate


- (void)vpkPreviewTouched:(VPKPreview *)preview image:(VPKImage*)image {
    /*
     invoking the VPKVeepViewer
     
     set the viewer's transitioning delegate to a custom transitioning object (or nil) to override supplied transition animations
     
     this code is all OPTIONAL - if you don't set the delegate on VPKPreview, presenting and dismissing behaviour occurs as a default
     
     */
    self.vpViewer = [VPKit viewerWithImage:image
                                  fromView:preview];
    self.vpViewer.delegate = self;
    self.vpViewer.modalPresentationStyle = UIModalPresentationOverFullScreen;
    [preview hideIcon];
    [self presentViewController:self.vpViewer animated:YES completion:nil];
}

#pragma mark - VPKViewController delegate

/*
    delegate methods for VPKVeepViewer
 
 */

- (void)veepViewer:(VPKVeepViewer *)viewer didFinishViewingWithInfo:(NSDictionary *)info {
    VPKPublicVeep* pVeep = info[@"veep"];
    NSLog(@"%s %@",__func__, pVeep);
    [self dismissViewControllerAnimated:YES completion:^{
        [self.vpkPreview showIcon];
    }];
}

- (void)veepViewerDidCancel:(VPKVeepViewer *)viewer {
    NSLog(@"%s",__func__);
    [self dismissViewControllerAnimated:YES completion:nil];
}

/*
    delegate mathods for VPKVeepEditor
 
 */

- (void)veepEditor:(VPKVeepEditor *)editor didPublishVeep:(NSString *)veepID  {
    [VPKit requestVeep:veepID completionBlock:
     ^(VPKPublicVeep * _Nullable veep, NSError * _Nullable error) {
         NSLog(@"%@",veep);
     }];
    NSLog(@"%s %@",__func__,veepID);
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)veepEditorDidCancel:(VPKVeepEditor *)editor {
    NSLog(@"%s",__func__);
    [self dismissViewControllerAnimated:YES completion:nil];
}


#pragma mark - error handling

/**
 Error handling

 
 @param notification NSNotification with userInfo dictionary containing a single item: the NSError at userInfo[VPKErrorKey].
 
 
 @discussion
 
 To propagate error notifications to the host app, call:
 [VPKit setForwardErrorNotifications:YES]  
 
 if notifications are not forwarded, they are handled by the SDK with user-facing alerts where appropriate.
 
 */

- (void)errorReceived:(NSNotification*)notification {
    NSError* error = notification.userInfo[VPKErrorKey];
    if (!error) return;
    [UIAlertController vpk_presentAlertWithError:error];
}

- (void)startListening {
    SEL selector = NSSelectorFromString(VPKErrorNotification);
    [NSNotificationCenter.defaultCenter addObserver:self
                                           selector:selector
                                               name:VPKErrorNotification
                                             object:nil];
}

- (void)stopListening {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

@end

File: VPKitDemoSwift/VPKitDemoSwift/ViewController.swift
//
//  ViewController.swift
//  VPKitDemoSwift
//
//  Created by jonathan on 09/11/2016.
//  Copyright © 2016 Veepio. All rights reserved.
//

import UIKit
import VPKit

class ViewController:
    UIViewController
    , VPKVeepEditorDelegate
    , VPKVeepViewerDelegate
    , VPKPreviewDelegate
{
    
    //MARK: - properties
    
    @IBOutlet var titleLabel: UILabel!
    @IBOutlet var consumeLabel: UILabel!
    @IBOutlet var createLabel: UILabel!
    @IBOutlet var vPreview: VPKPreview!
    @IBOutlet var ePreview: VPKPreview!
    
    
    
    //MARK: - viewController lifecycle
    
    required init?(coder aDecoder: NSCoder)
    {
        super.init(coder: aDecoder)
        //fatalError("init(coder:) has not been implemented")
    }

    override func viewDidLoad()
    {
        super.viewDidLoad()
        VPKit.setProduction(false)
        VPKit.setEmail("fake@email.com");
        self.configureViewerWithTestImage()
        self.configureEditor()
        self.configureConstraints()
        
    }
    
    

    //MARK: - configuration
    
    func configureViewerWithTestImage()
    {
        guard let image = UIImage.init(named: "KrispyGlas") else {return}
        let previewImage: VPKImage = VPKImage(image: image, veepID:"560")
        self.vPreview.image = previewImage
    }
    
    func configureViewerWithTestVideo()
    {
        guard let image = UIImage.init(named: "tomcruise") else {return}
        let previewImage: VPKImage = VPKImage(image: image, veepID:"680")
        self.vPreview.image = previewImage
    }
    
    func configureEditor()
    {
        guard let image = UIImage.init(named: "stock_photo") else {return}
        let previewImage: VPKImage = VPKImage(image: image, veepID:nil)
        self.ePreview.image = previewImage
        /*
         for the editor example, we'll set an optional delegate.
         */
        self.ePreview.delegate = self
    }
    
    
    
    
    //MARK: -  VPKPreview delegate
    
    func vpkPreviewTouched(_ preview:VPKPreview, image:VPKImage)
    {
        if preview.isEqual(self.vPreview) {
            self.invokeViewer(image: image, fromView: preview)
        } else {
            self.invokeEditor(image: image, fromView: preview)
        }
    }

    func invokeViewer(image:VPKImage, fromView view:UIView)
    {
        
        /*
         invoking the VPKVeepViewer
         
         set the viewer's transitioning delegate to a custom transitioning object (or nil) to override supplied transition animations
         
         this code is all OPTIONAL - if you don't set the delegate on VPKPreview, presenting and dismissing behaviour occurs as a default
         */
        
        
        guard let vpViewer = VPKit.viewer(with: image, from: view) else { return }
        vpViewer.delegate = self
        vpViewer.modalPresentationStyle = UIModalPresentationStyle.overFullScreen
        self.present(vpViewer, animated: true, completion: nil)
    }
    
    func invokeEditor(image:VPKImage, fromView view:UIView)
    {
        
        
        /*
         
         invoking the VPKVeepEditor
         
         
         set the editor's transitioning delegate to a custom transitioning object (or nil) to override supplied transition animations
         
         this code is all OPTIONAL - if you don't set the delegate on VPKPreview, presenting and dismissing behaviour occurs as a default. However, at some point before invoking the editor, user authentication needs to be dealt with.
         
         */
        
        
        VPKit.authenticate(withEmail: "public_test_2_4_6@example.com")
        {
            [weak self] (success, responseCode, error) in
            
            /*
             
             Authentication
             
             We authenticate before invoking the editor, as veep creation requires an authenticated user.
             
             Authenticated users are weak (no password) or strong (password-protected)
             
             If responseCode to a weak login attempt is 401 (unauthorised), we can make a similar call for strong authentication with a password.
             
             User account admin can be an implementation detail in the host app or the Veepio Developer control panel.
             
             */
            
            if (success) {
                guard let vpEditor = VPKit.editor(with: image, from: view) else { return }
                vpEditor.delegate = self
                vpEditor.modalPresentationStyle = UIModalPresentationStyle.overFullScreen
                self?.present(vpEditor, animated: true, completion: nil)

            } else {
                print("\(error)")
            }
        }
        
        
    }
    
    



    
    
    
    //MARK: - VPKViewController delegate functions
    
    /*
     delegate methods for VPKVeepViewer
     
     */
    
    func veepViewer(_ viewer: VPKVeepViewer, didFinishViewingWithInfo info: [AnyHashable : Any]?)
    {
        
        // the veep object can be read from info["veep"] if required

        self.dismiss(animated: true, completion:  {
            self.vPreview?.showIcon(true)
        })
    }
    
    func veepViewerDidCancel(_ viewer: VPKVeepViewer) {
        self.dismiss(animated: true, completion: nil)
    }
    
    /*
     delegate mathods for VPKVeepEditor
     
     */
    
    func veepEditor(_ editor: VPKVeepEditor, didPublishVeep veepID: String)
    {
        /*
         demo method to get a publicVeep object from a veepID
         */
        VPKit.requestVeep(veepID) { (veep, error) in
            if ((veep) != nil){
                print("\(veep)")
            } else {
                print("\(error)")
            }
        }
        self.dismiss(animated: true, completion: nil)
    }
    
    func veepEditorDidCancel(_ editor: VPKVeepEditor)
    {
        self.dismiss(animated: true, completion: nil)
    }
    
    
    //MARK: - layout boilerplate
    
    
    func configureConstraints()
    {
        let views:[String:UIView] =  [
            "titleLabel":self.titleLabel!
            ,"consumeLabel":self.consumeLabel!
            ,"createLabel":self.createLabel!
            ,"vPreview":self.vPreview!
            ,"ePreview":self.ePreview!
            
        ]
        for (_,view) in views {
            view.translatesAutoresizingMaskIntoConstraints = false
        }
        
        let vMargin = 50;
        let formats =
            [
                "H:|-(vMargin)-[titleLabel]-(vMargin)-|",
                "H:|-(vMargin)-[vPreview]-(vMargin)-|",
                "H:|-(vMargin)-[consumeLabel]-(vMargin)-|",
                "H:|-(vMargin)-[ePreview]-(vMargin)-|",
                "H:|-(vMargin)-[createLabel]-(vMargin)-|",
                "V:|-(38)-[titleLabel]-(18)-[vPreview]-(4)-[consumeLabel]-(18)-[ePreview]-(4)-[createLabel]"
        ];
        
        for format in formats {
            self.view.addConstraints(
                NSLayoutConstraint.constraints(withVisualFormat: format,
                                               options:NSLayoutFormatOptions(rawValue:0),
                                               metrics: ["vMargin":vMargin],
                                               views: views)
            )
        }
        
        guard let vImage = self.vPreview?.image else {return}
        var size = vImage.size;
        self.alignHeightToWidth(view:self.vPreview!, ratio: size.height/size.width);
        guard let eImage = self.ePreview?.image else {return}
        size = eImage.size;
        self.alignHeightToWidth(view:self.ePreview!, ratio: size.height/size.width);
        
    }
    
    func alignHeightToWidth (view:UIView, ratio:CGFloat)
    {
        view.superview?.addConstraint(
            NSLayoutConstraint.init(
                item: view,
                attribute: NSLayoutAttribute.height,
                relatedBy: NSLayoutRelation.equal,
                toItem: view,
                attribute: NSLayoutAttribute.width,
                multiplier: ratio, constant: 0)
        )
    }
    

}

